File: 297-serialize-and-deserialize-binary-tree/297-serialize-and-deserialize-binary-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */

import scala.collection.mutable._

class Codec {
    // Encodes a list of strings to a single string.
    def serialize(root: TreeNode): String = {
        val res = serializer(root)
        res
    }
    
    private def serializer(root: TreeNode): String = {
        if (root == null) { "" } 
        else { root.value + "," + serializer(root.left) + "," + serializer(root.right) }
    }
    
    // Decodes a single string to a list of strings.
    def deserialize(data: String): TreeNode = {
        deserializer(data.split(","), 0)._1
    }

    private def deserializer(data: Array[String], s: Int): (TreeNode, Int) = {
        if (s >= data.length || data(s).isEmpty()) {
            (null, s)
        } else {
            val value = Integer.parseInt(data(s))
            val node = TreeNode(value)
            val (lc, ls) = deserializer(data, s + 1)
            val (rc, rs) = deserializer(data, ls + 1)
            node.left = lc
            node.right = rc
            
            (node, rs)
        }
    }
}


/**
 * Your Codec object will be instantiated and called as such:
 * var ser = new Codec()
 * var deser = new Codec()
 * val s = ser.serialize(root)
 * val ans = deser.deserialize(s)
 */

File: 297-serialize-and-deserialize-binary-tree/297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "";
        return root.val + "," + serialize(root.left) + "," + serialize(root.right);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] arr = data.split(",");
        return helper(arr, 0).getKey();
    }
    
    Pair<TreeNode, Integer> helper(String[] arr, int s) {
        if (s >= arr.length || arr[s].isEmpty()) return new Pair(null, s + 1);
        
        TreeNode node = new TreeNode(Integer.valueOf(arr[s]));
        
        Pair<TreeNode, Integer> lp = helper(arr, s + 1);
        Pair<TreeNode, Integer> rp = helper(arr, lp.getValue());
        
        node.left = lp.getKey();
        node.right = rp.getKey();
        
        return new Pair(node, rp.getValue());
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));